BitFaster.Caching/Lfu/ConcurrentLfuCore.cs

```diff
@@ -277,686 +277,707 @@ private bool TryGetImpl(K key, [MaybeNullWhen(false)] out V value)
-            sb.Append("] Protected [");
-            sb.Append(string.Join(",", this.protectedLru.Select(n => n.Key.ToString())));
-            sb.Append("] Probation [");
-            sb.Append(string.Join(",", this.probationLru.Select(n => n.Key.ToString())));
-            sb.Append(']');
-
-            return sb.ToString();
-        }
-#endif
-    }
-
-    // Explicit layout cannot be a generic class member
-    [StructLayout(LayoutKind.Explicit, Size = 2 * Padding.CACHE_LINE_SIZE)]
-    internal struct PaddedInt
-    {
-        [FieldOffset(1 * Padding.CACHE_LINE_SIZE)] public int Value;
-    }
-}
+            }
+        }
+
+        public bool TryRemove(KeyValuePair<K, V> item)
+        {
+            if (this.dictionary.TryGetValue(item.Key, out var node))
+            {
+                if (EqualityComparer<V>.Default.Equals(node.Value, item.Value))
+                {
+                    var kvp = new KeyValuePair<K, N>(item.Key, node);
+
+#if NET6_0_OR_GREATER
+                    if (this.dictionary.TryRemove(kvp))
+#else
+                    // https://devblogs.microsoft.com/pfxteam/little-known-gems-atomic-conditional-removals-from-concurrentdictionary/
+                    if (((ICollection<KeyValuePair<K, N>>)this.dictionary).Remove(kvp))
+#endif
+                    {
+                        node.WasRemoved = true;
+                        AfterWrite(node);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public bool TryRemove(K key, [MaybeNullWhen(false)] out V value)
+        {
+            if (this.dictionary.TryRemove(key, out var node))
+            {
+                node.WasRemoved = true;
+                AfterWrite(node);
+                value = node.Value;
+                return true;
+            }
+
+            value = default;
+            return false;
+        }
+
+        public bool TryRemove(K key)
+        {
+            return this.TryRemove(key, out var _);
+        }
+
+        public bool TryUpdate(K key, V value)
+        {
+            if (this.dictionary.TryGetValue(key, out var node))
+            {
+                if (TypeProps<V>.IsWriteAtomic)
+                {
+                    node.Value = value;
+                    
+                    // It's ok for this to be lossy, since the node is already tracked
+                    // and we will just lose ordering/hit count, but not orphan the node.
+                    this.writeBuffer.TryAdd(node);
+                    TryScheduleDrain();
+                }
+                else
+                { 
+                    var newNode = policy.Create(key, value);
+                    if (this.dictionary.TryUpdate(key, newNode, node))
+                    { 
+                        // This can't be lossy, because we need to attach the new node.
+                        node.WasRemoved = true;

```

Comment:
There are 2 issues with time based expiry:

1. `ExpireAfterPolicy.OnWrite` will interpret this as a new node, because `node.GetNextInTimeOrder() == null`. GetExpireAfterUpdate will not be called.
2. `ExpireAfterPolicy.Create `will also call `GetExpireAfterCreate`.